src/subdomains/core/fee/fee.service.ts

```diff
@@ -0,0 +1,230 @@
+import { IsNull } from 'typeorm';
+import { CreateFeeDto } from './dto/create-fee.dto';
+import { Fee, FeeType } from './fee.entity';
+import { FeeRepository } from './fee.repository';
+
+export interface UserFeeRequest extends FeeRequestBase {
+  userData: UserData;
+}
+
+export interface FeeRequest extends FeeRequestBase {
+  accountType: AccountType;
+}
+
+export interface FeeRequestBase {
+  direction?: FeeDirectionType;
+  asset?: Asset;
+  txVolume?: number;
+}
+
+@Injectable()
+export class FeeService {
+  private readonly logger = new DfxLogger(FeeService);
+
+  constructor(
+    private readonly feeRepo: FeeRepository,
+    private readonly assetService: AssetService,
+    private readonly userDataService: UserDataService,
+  ) {}
+
+  async createFee(dto: CreateFeeDto): Promise<Fee> {
+    // check if exists
+    const existing = await this.feeRepo.findOne({
+      where: {
+        label: dto.label,
+        direction: dto.direction,
+      },
+    });
+
+    if (existing) throw new BadRequestException('Fee already created');
+    if (dto.type !== FeeType.DISCOUNT && dto.createDiscountCode)
+      throw new BadRequestException('Only discount Fees can have a discountCode');
+    if (dto.type === FeeType.BASE && !dto.accountType)
+      throw new BadRequestException('Base Fees must have an accountType');
+
+    // create the entity
+    const buy = this.feeRepo.create(dto);
+
+    if (dto.assetIds) {
+      const assets = [];
+
+      for (const assetId of dto.assetIds) {
+        const asset = await this.assetService.getAssetById(Number.parseInt(assetId));
+        if (!asset) throw new NotFoundException(`Asset with id ${assetId} not found`);
+        assets.push(asset.id);
+      }
+      buy.assets = assets.join(';');
+    }
+
+    if (dto.createDiscountCode) {
+      // create hash
+      const hash = Util.createHash(buy.label + buy.type).toUpperCase();
+      buy.discountCode = `${hash.slice(0, 4)}-${hash.slice(4, 8)}-${hash.slice(8, 12)}`;
+    }
+
+    // save
+    return this.feeRepo.save(buy);
+  }
+
+  async addDiscountCodeUser(userData: UserData, discountCode: string): Promise<void> {
+    const fee = await this.getFeeByDiscountCode(discountCode);
+    if (this.isExpiredFee(fee, { accountType: userData.accountType }))
+      throw new BadRequestException('Discount code is expired');
+
+    if (fee.maxUsages && (await this.hasMaxUsageExceeded(fee)))
+      throw new BadRequestException('Max usages for discount code taken');
+
+    await this.userDataService.addDiscountCode(userData, fee.id.toString());
+  }
+
+  async addFeeInternal(userData: UserData, feeId: number): Promise<void> {
+    const fee = await this.feeRepo.findOneBy({ id: feeId });
+    if (this.isExpiredFee(fee, { accountType: userData.accountType }))
+      throw new BadRequestException('Discount code is expired');
+
+    if (fee.maxUsages && (await this.hasMaxUsageExceeded(fee)))
+      throw new BadRequestException('Max usages for discount code taken');
```

Comment:
This looks the same than lines 80-84. Maybe we can add a method `verifyFee` or something?